src/misc/fs_utils.cpp

```diff
@@ -0,0 +1,46 @@
+/*
+ *  SPDX-License-Identifier: GPL-2.0-or-later
+ *
+ *  Copyright (C) 2022-2022  The DOSBox Staging Team
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License along
+ *  with this program; if not, write to the Free Software Foundation, Inc.,
+ *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.
+ */
+
+#include "fs_utils.h"
+
+#include "std_filesystem.h"
+
+std_fs::path maybe_shorten_path(const std_fs::path &orig_path) noexcept
+{
+	std::error_code abs_ec = {};
+	std::error_code rel_ec = {};
+
+	auto abs_path = std_fs::absolute(orig_path, abs_ec);
+	auto rel_path = std_fs::relative(orig_path, std_fs::current_path(), rel_ec);
+
+	auto smaller_of = [](const std_fs::path &original,
+	                     const std_fs::path &candidate,
+	                     const std::error_code &candidate_ec) {
+		if (candidate_ec)
+			return original;
+
+		const auto original_length  = original.string().length();
+		const auto candidate_length = candidate.string().length();
+
+		return original_length < candidate_length ? original : candidate;
+	};
+
+	return smaller_of(smaller_of(orig_path, rel_path, rel_ec), abs_path, abs_ec);
+}
```

Comment:
So it the idea here to simplify the path to its shortest possible representation? E.g. `./foo` or `./foo/subdir/../subdir/..` would both result in just `foo/`, so the shortest relative path from the current path? It's also not clear whether this function will always return relative paths?

If this function can convert relative paths into absolute ones, or vice versa, I'm not quite sure I like that. I think it would be better to turn this into something called `simplify_path` that would work like this:

- for relative paths (e.g. turns `./foo/bar/..` into just `foo/`)
- for absolute paths (e.g. turns `/foo/bar/../bar/qux/..` into `/foo/bar/`)

So absolute paths stay absolute paths, relative paths stay relative paths, they just get simplified.

What about the trailing path-separator character, would that be always included at the end if the path ends with a directory?

Reply:
When looking up our built-in resources, which start locally but then look relative to the executable's binary, and finally out in common system areas (which is where package maintainers might put them), the paths we get back from `std::filesystem::path` are almost always absolute.

The problem with that is they are huge.. and often wrap in the DOS console (which looks terrible).  So the problem I'm trying to solve is to always present the shortest form, whatever that might take.

I like the idea:

> something called simplify_path that would work like this:
> for relative paths (e.g. turns ./foo/bar/.. into just foo/)
> for absolute paths (e.g. turns /foo/bar/../bar/qux/.. into /foo/bar/)


As we move to std::filesystem and use the automated resource lookup, the paths we get are almost always absolute - and most don't have superflous /../ stuff in them that need to be baked out.

So this idea:

> So absolute paths stay absolute paths, relative paths stay relative paths, they just get simplified.

Would always go down the absolute route - and we would be give back a huge wrapping path.

Reply:
> for relative paths (e.g. turns ./foo/bar/.. into just foo/)
> for absolute paths (e.g. turns /foo/bar/../bar/qux/.. into /foo/bar/)

We can bake out any path indirection and `../` filler using [`canonical(..)`](https://en.cppreference.com/w/cpp/filesystem/canonical), so I'll add that as a third alternate for the function to consider.

Will also rename to `simplify_path()`.

Reply:
It's working pretty nice .. can feed it crazy stuff like:

`"./tests123/../valid/tests456///1/..//2/../3/../..";`

And it'll bake it down to:

`"valid/"`

Reply:
Thanks for the explanation, that sounds like a reasonable approach. Adding a comment about the included trailing slash for directories wouldn't hurt, though.